allennlp/data/dataset_readers/dataset_reader.py

```diff
@@ -335,5 +289,122 @@ def deserialize_instance(self, string: str) -> Instance:
+            return (transform(x) for x in islice)
+        return islice
+
+    def _instance_iterator(self, file_path: str) -> Iterable[Instance]:
+        cache_file: Optional[str] = None
+        if self._cache_directory:
+            cache_file = self._get_cache_location_for_file_path(file_path)
+
+        if cache_file is not None:
+            # Try to acquire lock on cache file.
+            cache_file_lock = FileLock(cache_file + ".lock", timeout=self.CACHE_FILE_LOCK_TIMEOUT)
+            try:
+                cache_file_lock.acquire()
+            except Timeout:
+                # Couldn't acquire the lock within the timeout specified, so we'll
+                # just log a warning and ignore the cache.
+                logger.warning(
+                    "Failed to acquire lock on dataset cache file within {}s",
+                    self.CACHE_FILE_LOCK_TIMEOUT,
+                )
+                cache_file = None
+
+        if cache_file is not None:
+            # Try to acquire lock on cache file.
+            cache_file_lock = FileLock(cache_file + ".lock", timeout=self.CACHE_FILE_LOCK_TIMEOUT)
+            try:
+                cache_file_lock.acquire()
+            except Timeout:
+                # Couldn't acquire the lock within the timeout specified, so we'll
+                # just log a warning and ignore the cache.
+                logger.warning(
+                    "Failed to acquire lock on dataset cache file within {}s",
+                    self.CACHE_FILE_LOCK_TIMEOUT,
+                )
+                cache_file = None
+
+        if cache_file is not None and os.path.exists(cache_file):
+            # If the file already exists and we've acquired the lock then it must
+            # be up-to-date. Therefore we can release the lock and start reading from it.
+            cache_file_lock.release()
+            logger.info("Reading instances from cache %s", cache_file)
+            with open(cache_file) as data_file:
+                for instance in self._multi_worker_islice(
+                    data_file, transform=self.deserialize_instance
+                ):
+                    yield instance
+        elif cache_file is not None:
+            # We've acquired the lock but the cache file doesn't exist. So we'll try
+            # writing to it.
+            if util.is_distributed() or (get_worker_info() and get_worker_info().num_workers):
+                # But we can't write to the cache if there's more than one process loading
+                # instances since each worker only receives a partial share of the instances.
+                logger.warning(
+                    "Can't cache data instances when there are multiple processes loading data"
+                )
+                # Release the lock, then just read the instances as normal.
+                cache_file_lock.release()
+                for instance in self._multi_worker_islice(self._read(file_path)):
+                    yield instance
```

Comment:
I think you can just say `yield from self._multi_worker_islice(...)`, or even `return self._multi_worker_islice(...)`. No loop necessary.

Reply:
I like `yield from ..`. But I just found out that putting a `return` statement in a generator does some very strange things.

For example:

```
def foo():
    return (x for x in [1, 2])
    yield "hi"

list(foo())
```

The result is just an empty list 😕

Reply:
Oh, interesting. So you can either `return my_iterator` or `yield from my_iterator`, but you better be consistent about it across different code paths?

Like this:
```
In [1]: def f():
   ...:     if True:
   ...:         return (x for x in [1, 2])
   ...:     else:
   ...:         yield "hi"
   ...:

In [2]: list(f())
Out[2]: []
```

That's ... something.

Reply:
Python encouraging people to shoot themselves in the foot.